src/Microsoft.ML.LightGBM/LightGbmTrainerBase.cs

```diff
@@ -0,0 +1,883 @@
+            }
+        }
+
+        private void GetFeatureValueSparse(IChannel ch, FloatLabelCursor cursor, CategoricalMetaData catMetaData, IRandom rand, out int[] indices, out float[] featureValues, out int cnt)
+        {
+            if (catMetaData.CategoricalBoudaries != null)
+            {
+                List<int> featureIndices = new List<int>();
+                List<float> values = new List<float>();
+                int lastIdx = -1;
+                int nhot = 0;
+                for (int i = 0; i < cursor.Features.Count; ++i)
+                {
+                    float fv = cursor.Features.Values[i];
+                    int colIdx = cursor.Features.Indices[i];
+                    int newColIdx = catMetaData.OnehotIndices[colIdx];
+                    if (catMetaData.IsCategoricalFeature[newColIdx])
+                        fv = catMetaData.OnehotBias[colIdx] + 1;
+                    if (newColIdx != lastIdx)
+                    {
+                        featureIndices.Push(newColIdx);
+                        values.Push(fv);
+                        nhot = 1;
+                    }
+                    else
+                    {
+                        // Multi-hot.
+                        ++nhot;
+                        var prob = rand.NextSingle();
+                        if (prob < 1.0f / nhot)
+                            values[values.Count() - 1] = fv;
+                    }
+                    lastIdx = newColIdx;
+                }
+                indices = featureIndices.ToArray();
+                featureValues = values.ToArray();
+                cnt = featureIndices.Count();
+            }
+            else
+            {
+                indices = cursor.Features.Indices;
+                featureValues = cursor.Features.Values;
+                cnt = cursor.Features.Count;
+            }
+        }
+
+        /// <summary>
+        /// Create a dataset from the sampling data.
+        /// </summary>
+        private void CreateDatasetFromSamplingData(IChannel ch, FloatLabelCursor.Factory factory,
+            int numRow, string param, float[] labels, float[] weights, int[] groups, CategoricalMetaData catMetaData,
+            out Dataset dataset)
+        {
+            Host.AssertValue(ch);
+
+            int numSampleRow = GetNumSampleRow(numRow, FeatureCount);
+
+            var rand = Host.Rand;
+            double averageStep = (double)numRow / numSampleRow;
+            int totalIdx = 0;
+            int sampleIdx = 0;
+            double density = DetectDensity(factory);
+
+            double[][] sampleValuePerColumn = new double[catMetaData.NumCol][];
+            int[][] sampleIndicesPerColumn = new int[catMetaData.NumCol][];
+            int[] nonZeroCntPerColumn = new int[catMetaData.NumCol];
+            int estimateNonZeroCnt = (int)(numSampleRow * density);
+            estimateNonZeroCnt = Math.Max(1, estimateNonZeroCnt);
+            Parallel.For(0, catMetaData.NumCol, i =>
```

Comment:
Here's another case where `Parallel.For` is probably slowing the operation down.